component/pyroscope/java/loop.go

```diff
@@ -0,0 +1,253 @@
+//go:build linux && (amd64 || arm64)
+
+package java
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/agent/component/discovery"
+	"github.com/grafana/agent/component/pyroscope"
+	"github.com/grafana/agent/component/pyroscope/java/asprof"
+	"github.com/grafana/agent/pkg/flow/logging/level"
+	jfrpprof "github.com/grafana/jfr-parser/pprof"
+	jfrpprofPyroscope "github.com/grafana/jfr-parser/pprof/pyroscope"
+	"github.com/prometheus/prometheus/model/labels"
+	gopsutil "github.com/shirou/gopsutil/v3/process"
+)
+
+const spyName = "grafana-agent.java"
+
+type profilingLoop struct {
+	logger log.Logger
+	output *pyroscope.Fanout
+
+	cfg       ProfilingConfig
+	wg        sync.WaitGroup
+	mutex     sync.Mutex
+	pid       int
+	target    discovery.Target
+	cancel    context.CancelFunc
+	error     error
+	dist      *asprof.Distribution
+	jfrFile   string
+	startTime time.Time
+	profiler  *asprof.Profiler
+}
+
+func newProfilingLoop(pid int, target discovery.Target, logger log.Logger, profiler *asprof.Profiler, output *pyroscope.Fanout, cfg ProfilingConfig) *profilingLoop {
+	ctx, cancel := context.WithCancel(context.Background())
+	dist, err := profiler.DistributionForProcess(pid)
+	p := &profilingLoop{
+		logger:   log.With(logger, "pid", pid),
+		output:   output,
+		pid:      pid,
+		target:   target,
+		cancel:   cancel,
+		dist:     dist,
+		jfrFile:  fmt.Sprintf("/tmp/asprof-%d-%d.jfr", os.Getpid(), pid),
+		cfg:      cfg,
+		profiler: profiler,
+	}
+	_ = level.Debug(p.logger).Log("msg", "new process", "target", fmt.Sprintf("%+v", target))
+
+	if err != nil {
+		p.onError(fmt.Errorf("failed to select dist for pid %d: %w", pid, err))
+		return p
+	}
+
+	p.wg.Add(1)
+	go func() {
+		defer p.wg.Done()
+		p.loop(ctx)
+	}()
+	return p
+}
+
+func (p *profilingLoop) loop(ctx context.Context) {
+	if err := p.profiler.CopyLib(p.dist, p.pid); err != nil {
+		p.onError(fmt.Errorf("failed to copy libasyncProfiler.so: %w", err))
+		return
+	}
+	defer p.stop()
+	sleep := func() bool {
+		timer := time.NewTimer(p.interval())
+		select {
+		case <-timer.C:
+			return false
+		case <-ctx.Done():
+			return true
+		}
+	}
+	for {
+		err := p.start()
+		if err != nil {
+			//  could happen when agent restarted - [ERROR] Profiler already started\n
+			alive := p.onError(fmt.Errorf("failed to start: %w", err))
+			if !alive {
+				return
+			}
+		}
+		done := sleep()
```

Comment:
you actually don't need done here. you could just do if ctx.Err() != nil { return }